File: ignition/iot_filter.py
from datetime import datetime

def herbicide_filter(di):
    if di["herbicide?"]:
        return di["spray herbicide"]
    return None

def insecticide_filter(di):
    if di["insecticide?"]:
        return di["spray insecticide"]
    return None

def fungicide_filter(di):
    if di["fungicide?"]:
        return di["spray fungicide"]
    return None

def fertilizer_filter(di):
    if di["fertilizer?"]:
        return di["spray fertilizer"]
    return None

def chemical_sum(arr):
    time_start = datetime.strptime(arr[0]["time"], "%m/%d/%y %H:%M:%S.%f")
    time_stop = time_start
    amount = 0
    for di in arr:
        amount += di["amount"]
        curr_time = datetime.strptime(di["time"], "%m/%d/%y %H:%M:%S.%f")
        if curr_time < time_start:
            time_start = curr_time
        elif curr_time > time_stop:
            time_stop = curr_time

    
    return {"id" : arr[0]["id"], 
            "pesticide": arr[0]["pesticide"],
            "total amount": amount,
            "start time": time_start.strftime("%m/%d/%y %H:%M:%S.%f"),
            "end time": time_stop.strftime("%m/%d/%y %H:%M:%S.%f")}

File: frontend/src/scripts/iotFilter.js
// Used ChatGPT to translate these from Python

function herbicideFilter(di) {
    if (di["herbicide?"]) {
        return di["spray herbicide"];
    }
    return null;
}

function insecticideFilter(di) {
    if (di["insecticide?"]) {
        return di["spray insecticide"];
    }
    return null;
}

function fungicideFilter(di) {
    if (di["fungicide?"]) {
        return di["spray fungicide"];
    }
    return null;
}

function fertilizerFilter(di) {
    if (di["fertilizer?"]) {
        return di["use fertilizer"];
    }
    return null;
}

function chemicalSum(arr) {
    let timeStart = new Date(arr[0]["time"]);
    let timeStop = timeStart;
    let amount = 0;
    let timesRun = arr.length;
    let cropCount = arr[0]["cropCount"];
    let cropType = arr[0]["cropType"];
    let location = arr[0]["location"];

    for (let di of arr) {
        amount += di["amount"];
        let currTime = new Date(di["time"]);
        if (currTime < timeStart) {
            timeStart = currTime;
        } else if (currTime > timeStop) {
            timeStop = currTime;
        }
    }

    return {
        "id": arr[0]["id"],
        "pesticide": arr[0]["chemical"],
        "total amount": amount,
        "times run": timesRun,
        "location": location,
        "crop count": cropCount,
        "crop type": cropType,
        "start time": timeStart, // ISO string formatted to match %m/%d/%y %H:%M:%S.%f
        "end time": timeStop
    };
}

export{
    herbicideFilter, insecticideFilter, fungicideFilter, fertilizerFilter, chemicalSum
}